system/lib/pthread/library_pthread.c

```diff
@@ -99,23 +94,21 @@ void emscripten_thread_sleep(double msecs) {
 
   emscripten_conditional_set_current_thread_status(
     EM_THREAD_STATUS_RUNNING, EM_THREAD_STATUS_SLEEPING);
-  now = emscripten_get_now();
-  while (now < target) {
+
+  double ms_to_sleep;
+  do {
     // Keep processing the main loop of the calling thread.
     __pthread_testcancel(); // pthreads spec: sleep is a cancellation point, so must test if this
                             // thread is cancelled during the sleep.
     emscripten_current_thread_process_queued_calls();
 
-    now = emscripten_get_now();
-    double ms_to_sleep = target - now;
-    if (ms_to_sleep > max_ms_slice_to_sleep) {
-      ms_to_sleep = max_ms_slice_to_sleep;
-    }
-    if (ms_to_sleep >= min_ms_slice_to_sleep) {
-      emscripten_futex_wait(&dummyZeroAddress, 0, ms_to_sleep);
-    }
-    now = emscripten_get_now();
-  };
+    ms_to_sleep = target - emscripten_get_now();
+    if (ms_to_sleep < min_ms_slice_to_sleep)
+      continue;
+
+    emscripten_futex_wait(&dummyZeroAddress, 0,
+      ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep);
```

Comment:
Here the code used to be of form
```c
emscripten_futex_wait(..., ms_to_sleep);
```
but it gets refactored to form
```c
emscripten_futex_wait(..., ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep);
```

Above in file `system/lib/libc/musl/src/thread/__wait.c` the code used to be

```c
emscripten_futex_wait(..., is_runtime_thread ? 1 : 100);
```
but it gets refactored to
```c
emscripten_futex_wait(..., max_ms_to_sleep);
```

The PR looks to me like it is massaging the code back and forth, I am not sure if I see the guiding idea here. (sorry!)

Reply:
Ah yes, it's a bit confusing. The difference between `system/lib/libc/musl/src/thread/__wait.c` and  `system/lib/pthread/library_pthread.c` is that the former needs to wait indefinitely in slices, so it's doing this:
```c
emscripten_futex_wait(..., max_ms_slice_to_sleep);
```
and the later must wait in slices for a certain time, so it does this:
```c
emscripten_futex_wait(..., MAX(ms_to_sleep, max_ms_slice_to_sleep));
```

I've renamed the `max_ms_to_sleep` variable to `max_ms_slice_to_sleep`  with commit 5b77bd909c717da9e11781a01e3d0a0a1b7c0039.

Reply:
I'd rather avoid the ternary operation here.  I think it makes the code easier to read if the calculation for how long to sleep is doing before the line where the sleep occurs.    It also make it easier for debugging since you have a name for the actual amount of time that is we are going to sleep for.

Don't you find "ms_to_sleep > max_ms_slice_to_sleep ? max_ms_slice_to_sleep : ms_to_sleep" in addition to need to function call split over two lines harder to read?

Reply:
You're right, those long variable names make it hard to read when being used within a ternary operator. Split over 2 lines with commit f84985c239f0a0a089248e42f1a3ccebe430d1cf to improve readability.